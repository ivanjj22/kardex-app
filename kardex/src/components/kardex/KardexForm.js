import React from "react";
import { useForm } from "../../hooks/useForm";
import url from "../../config";
import { useParams } from "react-router-dom";

export const KardexForm = ({ onInsertKardex}) => {
  let { id } = useParams();

  const [form, onChange, resetForm] = useForm({
    date: "",
    type: "",
    units: "",
    units_price: "",
    product_id: id,
  });

  const handleSubmit = async (e) => {
    e.preventDefault();
    if(!form.date || !form.type || !form.units || !form.units_price) {
      return false;
    }
    
    try {
      let config = {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(form),
      };
      await fetch(`${url}/movements`, config);
      onInsertKardex();
      resetForm();
    } catch (error) {
      console.error(error)
    }
  };

  return (
    <form className="w-full mt-5 mb-5" onSubmit={handleSubmit}>
      <div className="flex flex-wrap -mx-3 mb-6">
        <div className="w-full md:w-1/5 px-3 mb-6 md:mb-0">
          <label
            className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
            htmlFor="date"
          >
            Fecha
          </label>
          <input
            className="dateInput appearance-none block w-full bg-gray-200 text-gray-700 border rounded py-3 px-4 mb-3 leading-tight focus:outline-none focus:bg-white"
            name="date"
            type="date"
            placeholder="Date"
            onChange={onChange}
            value={form.date}
          />
        </div>
        <div className="w-full md:w-1/5 px-3">
          <label
            className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
            htmlFor="type"
          >
            Tipo
          </label>
          <select
            className="typeInput appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
            name="type"
            type="text"
            placeholder="Tipo"
            onChange={onChange}
            value={form.type}
          >
            <option value="">Seleccione</option>
            <option value="init">Inventario inicial</option>
            <option value="purchase">Compra</option>
            <option value="selling">Venta</option>
          </select>
        </div>
        <div className="w-full md:w-1/5 px-3">
          <label
            className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
            htmlFor="units"
          >
            Cantidad
          </label>
          <input
            className="unitsInput appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
            name="units"
            type="number"
            placeholder="Cantidad"
            onChange={onChange}
            value={form.units}
            min="1"
          />
        </div>
        <div className="w-full md:w-1/5 px-3">
          <label
            className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
            htmlFor="units_price"
          >
            Valor unitario
          </label>
          <input
            className="UnitsPriceInput appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
            name="units_price"
            type="number"
            placeholder="Valor unitario"
            onChange={onChange}
            value={form.units_price}
            min="1"
          />
        </div>
        <div className="w-full md:w-1/5 px-3 my-auto">
          <button className="btn btn-sm btn-success text-white">Guardar</button>
          <button className="btn btn-sm btn-error mx-2 text-white">
            Limpiar
          </button>
        </div>
      </div>
    </form>
  );
};
