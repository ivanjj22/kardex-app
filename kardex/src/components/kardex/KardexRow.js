import React from "react";

export const KardexRow = ({ id, date, detail, data_in, data_out, balance }) => {
  return (
    <tr key={id}>
      <th className="border px-8 py-4">{date}</th>
      <td className="border px-8 py-4">{detail}</td>
      <td className="border px-8 py-4">
        <table style={{ textAlign: "center" }} width="100%">
          <tbody>
            <tr>
              <td>{data_in.units}</td>
              <td>{data_in.units_price}</td>
              <td>{data_in.total}</td>
            </tr>
          </tbody>
        </table>
      </td>
      <td className="border px-8 py-4">
        <table style={{ textAlign: "center" }} width="100%">
          <tbody>
            <tr>
              <td>{data_out.units}</td>
              <td>{data_out.units_price}</td>
              <td>{data_out.total}</td>
            </tr>
          </tbody>
        </table>
      </td>
      <td className="border px-8 py-4 ">
        <table style={{ textAlign: "center" }} width="100%">
          <tbody>
            <tr>
              <td>{balance.units}</td>
              <td>{balance.units_price}</td>
              <td>{balance.total}</td>
            </tr>
          </tbody>
        </table>
      </td>
    </tr>
  );
};
