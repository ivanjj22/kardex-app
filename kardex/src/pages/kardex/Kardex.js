import React from 'react'

export const Kardex = () => {
    return (
        <div>
            
        </div>
    )
}
