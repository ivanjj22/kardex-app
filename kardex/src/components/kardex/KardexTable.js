import React, { useState, useEffect } from "react";
import { KardexRow } from "./KardexRow";
import url from "../../config";
import { useParams } from "react-router-dom";

export const KardexTable = ({ current }) => {
  const [data, setData] = useState([]);

  let { id } = useParams();

  useEffect(() => {
    loadInitial();
  }, [current]);

  async function loadInitial() {
    fetch(`${url}/movements?product_id=${id}`)
      .then((resp) => resp.json())
      .then((data) => {
        setData(data.data);
      })
      .catch(() => {});
  }

  if (!data) return null;

  return (
    <table className="shadow-lg bg-white w-full">
      <thead>
        <tr>
          <th className="bg-blue-100 border text-left px-8 py-4">Fecha</th>
          <th className="bg-blue-100 border text-left px-8 py-4">Detalle</th>
          <th className="bg-blue-100 border text-left px-8 py-4">
            <table style={{ textAlign: "center" }} width="100%">
              <tbody>
                <tr>
                  <td colSpan="3" className="titles">
                    Entradas
                  </td>
                </tr>
                <tr>
                  <td>Canti</td>
                  <td>Unit</td>
                  <td>Total</td>
                </tr>
              </tbody>
            </table>
          </th>
          <th className="bg-blue-100 border text-left px-8 py-4">
            <table style={{ textAlign: "center" }} width="100%">
              <tbody>
                <tr>
                  <td colSpan="3" className="titles">
                    Salidas
                  </td>
                </tr>
                <tr>
                  <td>Canti</td>
                  <td>Unit</td>
                  <td>Total</td>
                </tr>
              </tbody>
            </table>
          </th>
          <th className="bg-blue-100 border text-left px-8 py-4">
            <table style={{ textAlign: "center" }} width="100%">
              <tbody>
                <tr>
                  <td colSpan="3" className="titles">
                    Saldos
                  </td>
                </tr>
                <tr>
                  <td>Canti</td>
                  <td>Unit</td>
                  <td>Total</td>
                </tr>
              </tbody>
            </table>
          </th>
        </tr>
      </thead>
      <tbody>
        {data.map((movement) => (
          <KardexRow key={movement.id} {...movement} />
        ))}
      </tbody>
    </table>
  );
};
