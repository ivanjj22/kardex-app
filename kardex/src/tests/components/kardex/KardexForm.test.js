import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { KardexForm } from '../../../components/kardex/KardexForm';
import { act} from '@testing-library/react'

describe('Pruebas en <KardexForm />', () => {

    test('debe de mostrarse correctamente', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/products/a775c47c-9961-41db-8275-3ae9dfbg0e04']}>
                <Route 
                    path="/products/:id" 
                    component={ () => <KardexForm form={{}} /> } 
                />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();

    })

    test('debe de hacer submit al formulario', async () => {

        const onInsertKardex = jest.fn();

        const wrapper = mount(
            <MemoryRouter initialEntries={['/products/a775c47c-9961-41db-8275-3ae9dfbg0e04']}>
                <Route 
                    path="/products/:id" 
                    component={ () => <KardexForm onInsertKardex={onInsertKardex} /> } 
                />
            </MemoryRouter>
        );
        
        wrapper.find('.dateInput').simulate('change', {
            target: {
                name: 'date',
                value: '2021/05/23'
            }
        });

        wrapper.find('.typeInput').simulate('change', {
            target: {
                name: 'type',
                value: 'init'
            }
        });

        wrapper.find('.unitsInput').simulate('change', {
            target: {
                name: 'units',
                value: '1000'
            }
        });

        wrapper.find('.UnitsPriceInput').simulate('change', {
            target: {
                name: 'units_price',
                value: '8000'
            }
        });

        await act(async () => {
            await wrapper.find('form').props().onSubmit({ preventDefault: () => {} });
        });

        expect(onInsertKardex).toHaveBeenCalled();
    
    })

    
})
