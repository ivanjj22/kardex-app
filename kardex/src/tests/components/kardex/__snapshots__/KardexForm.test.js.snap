// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <KardexForm /> debe de mostrarse correctamente 1`] = `
<form
  className="w-full mt-5 mb-5"
  onSubmit={[Function]}
>
  <div
    className="flex flex-wrap -mx-3 mb-6"
  >
    <div
      className="w-full md:w-1/5 px-3 mb-6 md:mb-0"
    >
      <label
        className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
        htmlFor="date"
      >
        Fecha
      </label>
      <input
        className="dateInput appearance-none block w-full bg-gray-200 text-gray-700 border rounded py-3 px-4 mb-3 leading-tight focus:outline-none focus:bg-white"
        name="date"
        onChange={[Function]}
        placeholder="Date"
        type="date"
        value=""
      />
    </div>
    <div
      className="w-full md:w-1/5 px-3"
    >
      <label
        className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
        htmlFor="type"
      >
        Tipo
      </label>
      <select
        className="typeInput appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
        name="type"
        onChange={[Function]}
        placeholder="Tipo"
        type="text"
        value=""
      >
        <option
          value=""
        >
          Seleccione
        </option>
        <option
          value="init"
        >
          Inventario inicial
        </option>
        <option
          value="purchase"
        >
          Compra
        </option>
        <option
          value="selling"
        >
          Venta
        </option>
      </select>
    </div>
    <div
      className="w-full md:w-1/5 px-3"
    >
      <label
        className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
        htmlFor="units"
      >
        Cantidad
      </label>
      <input
        className="unitsInput appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
        min="1"
        name="units"
        onChange={[Function]}
        placeholder="Cantidad"
        type="number"
        value=""
      />
    </div>
    <div
      className="w-full md:w-1/5 px-3"
    >
      <label
        className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
        htmlFor="units_price"
      >
        Valor unitario
      </label>
      <input
        className="UnitsPriceInput appearance-none block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
        min="1"
        name="units_price"
        onChange={[Function]}
        placeholder="Valor unitario"
        type="number"
        value=""
      />
    </div>
    <div
      className="w-full md:w-1/5 px-3 my-auto"
    >
      <button
        className="btn btn-sm btn-success text-white"
      >
        Guardar
      </button>
      <button
        className="btn btn-sm btn-error mx-2 text-white"
      >
        Limpiar
      </button>
    </div>
  </div>
</form>
`;
